src/d/actor/d_a_kui.cpp

```diff
@@ -4,68 +4,541 @@
+            cXyz vec(REG8_F(1) + 800.f, REG8_F(2), REG8_F(3) + 100.0f);
+            cXyz posVec;
+            MtxPosition(&vec, &posVec);
+        
+            i_this->field_0x2EC = i_this->home.pos + posVec;
+        
+            if (i_this->field_0x2EA == 70) {
+                i_this->field_0x2E8 = 0;
+        
+                camera->mCamera.Reset(i_this->field_0x2F8, i_this->field_0x2EC);
+                camera->mCamera.Start();
+                camera->mCamera.SetTrimSize(0);
+        
+                dComIfGp_event_reset();
+        
+                dComIfGs_onSwitch(i_this->mSwitchNo, fopAcM_GetRoomNo(i_this));
+                mDoAud_seStart(JA_SE_READ_RIDDLE_1);
+            }
+            break;
+    }
+
+    if (((s8)i_this->field_0x2E8 != 0) && bVar2) {
+        camera->mCamera.Set(i_this->field_0x2F8, i_this->field_0x2EC);
+        JUTReport(0x19a, 0x1ae, "K SUB  COUNT  %d", i_this->field_0x2EA);
+        i_this->field_0x2EA++;
+    }
 }
 
 /* 00000920-000012E4       .text daKui_Execute__FP9kui_class */
-static BOOL daKui_Execute(kui_class*) {
-    /* Nonmatching */
+static BOOL daKui_Execute(kui_class* i_this) {
+    fopAc_ac_c* actor = i_this;
+    static s16 bure_xa_d[2] = {0x1000, 0xF000};
+
+    cXyz temp2;
+    cXyz temp;
+    daPy_py_c* player;
+    s16 target_x_angle;
+    Mtx local_mtx;
+
+    player = (daPy_py_c*)dComIfGp_getPlayer(0);
+
+    if (i_this->field_0x2A2 != 0) {
+        dr2_class* dragon_tail = search_dragontail(i_this);
+        btd_class* btd = search_btd(i_this);
+
+        if (dragon_tail != NULL && btd != NULL) {
+            if ((dragon_tail->field_0x4BA == 0 || dragon_tail->field_0x4BA >= 10) && btd->field_0x2e4 < 10) {
+                actor->current.pos = dragon_tail->field_0x3B0;
+                actor->current.angle = dragon_tail->current.angle;
+            } else {
+                actor->current.pos.set(0.0f, -10000.0f, 0.0f);
+            }
+        } else {
+            actor->current.pos.set(0.0f, -10000.0f, 0.0f);
+        }
+    }
+
+    if (i_this->type == 3) {
+        if (actor->health == 3) {
+            temp2 = player->getLeftHandPos() - actor->home.pos;
+            cMtx_YrotS(*calc_mtx, -player->shape_angle.y);
+
+            MtxPosition(&temp2, &temp);
+            temp.z *= REG0_F(1) + 1.0f;
+
+            if (REG0_S(0) == 0) {
+                target_x_angle = -cM_atan2s(temp.z, -temp.y);
+            } else {
+                target_x_angle = (s32) cM_atan2s(temp.z, -temp.y);
+            }
+            s16 unk = std::fabsf((REG0_F(2) + 3000.0f) * cM_ssin(actor->shape_angle.y));
+
+            s8 unk_flag = 0;
+            if (target_x_angle > unk) {
+                target_x_angle = unk;
+                unk_flag = 1;
+            } else if (target_x_angle < (s16)-unk) {
+                target_x_angle = -unk;
+                unk_flag = 2;
+            }
+
+            if (unk_flag != 0 && unk > 2000) {
+                s16* shorts_array = (&i_this->field_0x2DC);
```

Comment:
Does changing field_0x2DC's type to `s16[3]` not match in some way?

Reply:
I thought I tried it earlier but I must have had some other error. It works with a `s16[3]` now and is much cleaner, thanks!